src/client/api/rpc.c

```diff
@@ -122,6 +123,16 @@ query_cb(struct crt_proto_query_cb_info *cb_info)
 		nr_ranks = dc_mgmt_net_get_num_srv_ranks();
 		D_ASSERT(nr_ranks > 0);
 		rproto->rank_idx = (rproto->rank_idx + 1) % nr_ranks;
+
+		/** We made a full loop and found no engines alive */
+		if (rproto->start_rank_idx == rproto->rank_idx) {
```

Comment:
I see one case when this test could not prevent some infinite loop.
If `rank_idx` is initially equal to `dc_mgmt_net_get_num_srv_ranks()` and one or more server are excluded during the loop, then the loop could run infinitely.
However, if I am correct, I do not see how to easily deal with this issue.

Reply:
good point. i pushed another version now that instead uses num_retries_left which is set initially to nr_ranks, so that we avoid such scenario as well. This will also allow set arbitrary number of retries later easier instead of being stuck with 'tried all engines once'